pkg/costmodel/aggregation.go

```diff
@@ -0,0 +1,2000 @@
+		// values by closure to prevent reference-type looping bug.
+		// (see https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable)
+		for label, values := range labelValues {
+			ff := (func(l string, vs []string) FilterFunc {
+				return func(cd *CostData) (bool, string) {
+					ae := aggregateEnvironment(cd)
+					for _, v := range vs {
+						if v == "__unallocated__" { // Special case. __unallocated__ means return all pods without the attached label
+							if _, ok := cd.Labels[label]; !ok {
+								return true, ae
+							}
+						}
+						if cd.Labels[label] == v {
+							return true, ae
+						} else if strings.HasSuffix(v, "*") { // trigger wildcard prefix filtering
+							vTrim := strings.TrimSuffix(v, "*")
+							if strings.HasPrefix(cd.Labels[label], vTrim) {
+								return true, ae
+							}
+						}
+					}
+					return false, ae
+				}
+			})(label, values)
+			filterFuncs = append(filterFuncs, ff)
+		}
+	}
+
+	// clear cache prior to checking the cache so that a clearCache=true
+	// request always returns a freshly computed value
+	if clearCache {
+		a.AggregateCache.Flush()
+		a.CostDataCache.Flush()
+	}
+
+	cacheExpiry := a.GetCacheExpiration(window.Duration())
+	if noExpireCache {
+		cacheExpiry = cache.NoExpiration
+	}
+
+	// parametrize cache key by all request parameters
+	aggKey := GenerateAggKey(window, field, subfields, opts)
+
+	thanosOffset := time.Now().Add(-thanos.OffsetDuration())
+	if a.ThanosClient != nil && window.End().After(thanosOffset) {
+		log.Infof("ComputeAggregateCostModel: setting end time backwards to first present data")
+
+		// Apply offsets to both end and start times to maintain correct time range
+		deltaDuration := window.End().Sub(thanosOffset)
+		s := window.Start().Add(-1 * deltaDuration)
+		e := time.Now().Add(-thanos.OffsetDuration())
+		window.Set(&s, &e)
+	}
+
+	dur, off := window.ToDurationOffset()
+	key := fmt.Sprintf(`%s:%s:%fh:%t`, dur, off, resolution.Hours(), remoteEnabled)
+
+	// report message about which of the two caches hit. by default report a miss
+	cacheMessage := fmt.Sprintf("ComputeAggregateCostModel: L1 cache miss: %s L2 cache miss: %s", aggKey, key)
+
+	// check the cache for aggregated response; if cache is hit and not disabled, return response
+	if value, found := a.AggregateCache.Get(aggKey); found && !disableCache && !noCache {
+		result, ok := value.(map[string]*Aggregation)
+		if !ok {
+			// disable cache and recompute if type cast fails
+			log.Errorf("ComputeAggregateCostModel: caching error: failed to cast aggregate data to struct: %s", aggKey)
+			return a.ComputeAggregateCostModel(promClient, window, field, subfields, opts)
+		}
+		return result, fmt.Sprintf("aggregate cache hit: %s", aggKey), nil
+	}
+
+	if window.Hours() >= 1.0 {
+		// exclude the last window of the time frame to match Prometheus definitions of range, offset, and resolution
+		start := window.Start().Add(resolution)
+		window.Set(&start, window.End())
```

Comment:
We might want to revisit this and do so, even when `window.Hours() < 1.0` (because now we're saying that resolution will be 1m, and the same issue holds)